xde/src/xde.rs

```diff
@@ -856,6 +869,84 @@ fn set_xde_underlay(req: &SetXdeUnderlayReq) -> Result<NoResp, OpteError> {
     Ok(NoResp::default())
 }
 
+#[no_mangle]
+fn clear_xde_underlay() -> Result<NoResp, OpteError> {
+    let state = get_xde_state();
+    let mut underlay = state.underlay.lock();
+    if underlay.is_none() {
+        return Err(OpteError::System {
+            errno: ENOENT,
+            msg: "underlay not yet initialized".into(),
+        });
+    }
+    if unsafe { xde_devs.read().len() } > 0 {
+        return Err(OpteError::System {
+            errno: EBUSY,
+            msg: "underlay in use by attached ports".into(),
+        });
+    }
+
+    if let Some(underlay) = underlay.take() {
+        // There shouldn't be anymore refs to the underlay given we checked for
+        // 0 ports above.
+        let Ok(u1) = Arc::try_unwrap(underlay.u1) else {
+            return Err(OpteError::System {
+                errno: EBUSY,
+                msg: "underlay u1 has outstanding refs".into(),
+            });
+        };
+        let Ok(u2) = Arc::try_unwrap(underlay.u2) else {
+            return Err(OpteError::System {
+                errno: EBUSY,
+                msg: "underlay u2 has outstanding refs".into(),
+            });
+        };
+
+        for u in [u1, u2] {
+            // Clear all Rx paths
+            u.mch.clear_rx();
+
+            // We have a chain of refs here:
+            //  1. `MacPromiscHandle` holds a ref to `MacClientHandle`, and
+            //  2. `MacUnicastHandle` holds a ref to `MacClientHandle`, and
+            //  3. `MacClientHandle` holds a ref to `MacHandle`.
+            // We explicitly drop them in order here to ensure there are no
+            // outstanding refs.
+
+            // 1. Remove promisc and unicast callbacks
+            drop(u.mph);
+            drop(u.muh);
+
+            // 2. Remove MAC client handle
+            if Arc::strong_count(&u.mch) > 1 {
```

Comment:
This isn't obviously correct; there could be a race between checking `Arc::strong_count` and someone else cloning the `Arc` (maybe this race is impossible due to other factors, but it's not obvious here).

Using `Arc::into_inner` (and removing `drop(u.mch)` below) would be more obviously correct.

Reply:
I don't think that this race is at all impossible -- `xde_rx` has another strong handle to the MAC client handle here, and it could be called into just before we deregister the promisc handle. 

That gives two bad orderings between {detach (today), underlay clear} and `xde_rx`, which makes me a little nervous if the two happen concurrently: we either bail on cleaning up our extra clients and leak holds on the underlying NIC, or `xde_rx` increments the strong count on an invalid `Arc` (!!!).

I think the only place this should be a risk is dev environments today, but I'm going to have a think about how we can safely quiesce here. Removing all ports is not a safe precondition, as any underlay traffic (or MAC flow matches, in future) will still land in `xde_rx`.

Reply:
I spent a little time double checking promisc handle behaviour in illumos, and it does give us the guarantees we need. I've left a comment explaining its behaviour for the benefit of future readers. 😄